core/src/main/java/fr/sncf/osrd/speedcontroller/generators/ConstructionAllowanceGenerator.java

```diff
@@ -119,15 +101,21 @@ protected double getFirstGuess() {
                     (integrator) -> integrator.actionToTargetSpeed(directive, schedule.rollingStock, -1));
             speed = step.finalSpeed;
             location.updatePosition(schedule.rollingStock.length, step.positionDelta);
+            acceleratingPhase.put(location.getPathPosition(), stepSpeed);
         }
 
-        var initialSpeedControllers = new HashSet<>(maxSpeedControllers);
-        // scaled MapSpeedControllers only between the end of the braking phase and the beginning of the acceleration
-        var speedsEndingEarlier = getExpectedSpeeds(sim, schedule, initialSpeedControllers, TIME_STEP,
-                endBrakingPosition, location.getPathPosition(), initialSpeed);
-        speedsEndingEarlier.put(endBrakingPosition, targetSpeed); // add first point to the future MapSpeedController
-        res.add(new MapSpeedController(speedsEndingEarlier).scaled(scaleFactor));
-
+        var coastingLastPosition = coastingPhase.lastKey();
+        var accelerationFirstPosition = location.getPathPosition();
+        if (accelerationFirstPosition == coastingLastPosition)
+            throw new SimulationError("Construction margin value too high for such short distance");
+        if (coastingPhase.size() > 1) {
+            res.add(new CoastingSpeedController(initialPosition, coastingLastPosition));
+        }
+        var marecoSpeedControllers =
+                super.getSpeedControllers(schedule, targetSpeed, coastingLastPosition, accelerationFirstPosition);
+        for (SpeedController speedController : marecoSpeedControllers) {
+            res.add(speedController);
+        }
```

Comment:
```suggestion
        res.addAll(marecoSpeedControllers);
```

Reply:
Nice, i didn't know i could do this, thank you !